tests/integrationtests/tst_perfparser.cpp

```diff
@@ -286,6 +286,21 @@ private slots:
         }
     }
 
+    /* tests a perf file that has data with PERF_FORMAT_LOST attribute, see KDAB/hotspot#578 */
+    void testPerfFormatLost()
+    {
+        PerfParser parser(this);
+        QSignalSpy parsingFailedSpy(&parser, &PerfParser::parsingFailed);
+        QSignalSpy parsingFinishedSpy(&parser, &PerfParser::parsingFinished);
+
+        parser.startParseFile(QFINDTESTDATA("perf.data.PerfFormatLost"));
+
+        // CHECKME: hotspot-perfparser alone needs < 0.2 seconds to parse this file,
```

Comment:
this is outdated now, is it?

Reply:
I don't understand this message then. can you please rephrase?

what is "newar instant open"?

and is the test running with DEBUGINFOD_URLS maybe? have you profiled the test to see where it takes the time?

for me the test finishes in ~2s when my debuginfod cache isn't filled. This is probably bad - the tests should run without any DEBUGINFOD_URLS internally, as that wouldn't be portable anyways. I suggest we forcefully unset this in the `initTestCase`

Reply:
> what is "newar instant open"?

Sorry, typo "newly instant open" in hotspot, of course (minimal) faster when just calling hotspot-perfparser with that, but takes very long during `ctest -VV`.

`DEBUGINFOD_URLS` is a good idea, but it isn't set in that environment.

> I suggest we forcefully unset this in the `initTestCase`

sounds good, I'd add `qputenv("DEBUGINFOD_URLS", QByteArrayLiteral(""));` as first line there in separate commit, do you agree?

> have you profiled the test to see where it takes the time?

Not yet, can you give me a hint how to run a single test in this environment (other than temporarily dropping all other tests from the source)? `ctest -N` only lists 1 integrationtest: tst_perfparser.

Reply:
running through callgrind to get more information on the child processes shows that the one that takes time is indeed `/workspace/hotspot/build/lib/x86_64-linux-gnu/libexec/hotspot-perfparser --input /workspace/hotspot/tests/integrationtests/perf.data.PerfFormatLost --max-frames 1024 --extra /workspace/hotspot/tests/integrationtests/file_content --app /usr/bin`

The interesting part: it only takes that long because of `--app /usr/bin`, when I manually call it that way `time` says ~ 55 seconds, if I leave it away I'm down to < 2 seconds.

If the the app option is specified then **a lot** of time is spent in QDir::entryList within QT5 in [perfsymboltable.cpp `findInExtraPath()`](https://github.com/KDAB/perfparser/blob/ab14b6989d123af60f4c822e8ece8638945139c6/app/perfsymboltable.cpp#L107).

![grafik](https://github.com/KDAB/hotspot/assets/6699539/40ed3983-61ec-4b43-ae0b-75c97357ca24)


It _could_ be that you're running a libQT5 that has an improved version of that.

![grafik](https://github.com/KDAB/hotspot/assets/6699539/303f288c-a57e-47ec-9155-a5b2e8e0d9ea)

Reply:
... the reason for `--app` are my setting adjustments in `testFileContent()`, if I drop those the test runs under 2 seconds also under `ctest`.

Questions:
* Is there a way to "backup" the current settings (ExtraLibPaths / AppPath)? If yes then this should be done and reset on each test that adjust those, if the tests may be run in parallel, then we'd need to lock on the settings object.
* Is the `qputenv("DEBUGINFOD_URLS", QByteArrayLiteral(""));` the right thing to add in `initTestCase()`?
* Most interesting one: is there any option to improve perfparser's [`findInExtraPath()`](https://github.com/KDAB/perfparser/blob/ab14b6989d123af60f4c822e8ece8638945139c6/app/perfsymboltable.cpp#L107) and/or the QT internal handling of the used call?
Ideally we tackle that before adjusting the test case.

Reply:
I have an idea for an optimization here, caching the path entries found; in overall we should not have _that_ much different paths (and otherwise we still temporarily create a bunch of QStrings plus filesystem lookup with `const auto entries = absDir.entryList({}, QDir::Dirs | QDir::NoDotAndDotDot);`, so the overhead of a cache + lookup is neglible).

... and checking what the underlying issue is: There is a famous [recursive symlink in x11-common](https://unix.stackexchange.com/questions/222919/infinitely-nested-directories-within-usr-bin-x11) `lrwxrwxrwx 1 root root 1 Mar 25  2022 /usr/bin/X11 -> .`, so the iteration is done "more or less" in QFileInfo until MAX_PATH - for each library lookup. Of course that's expensive!

The cache should help here, too when the key for the QSet is resolved by `path.canonicalPath`.

Reply:
https://github.com/KDAB/perfparser/pull/34 solves the lookup time with the "bad" /usr/bin/X11 beneath appPath/extraPath; I'm adjusting the comments and push again.